src/test/scala/org/zalando/spearheads/innkeeper/api/ComplexRoutesJsonProtocolsSpec.scala

```diff
@@ -0,0 +1,354 @@
+          |      "match": "/hello-*",
+          |      "type": "REGEX"
+          |    },
+          |    "methodMatcher": "POST",
+          |    "headerMatchers": [{
+          |      "name": "X-Host",
+          |      "value": "www.*",
+          |      "type": "REGEX"
+          |    }, {
+          |      "name": "X-Port",
+          |      "value": "8080",
+          |      "type": "STRICT"
+          |    }]
+          |  },
+          |  "filters": [{
+          |    "name": "someFilter",
+          |    "args": ["Hello", 123]
+          |  }, {
+          |    "name": "someOtherFilter",
+          |    "args": ["Hello", 123, "World"]
+          |  }],
+          |  "endpoint": "https://www.endpoint.com:8080/endpoint"
+          |}""".stripMargin
+      }
+    }
+
+    it("should marshall a minimal NewComplexRoute") {
+      val routeJson = NewComplexRoute(
+        matcher = Matcher(
+          pathMatcher = Some(RegexPathMatcher("/hello-*"))
+        )
+      ).toJson
+
+      routeJson.prettyPrint should be {
+        """{
+          |  "matcher": {
+          |    "pathMatcher": {
+          |      "match": "/hello-*",
+          |      "type": "REGEX"
+          |    },
+          |    "headerMatchers": []
+          |  },
+          |  "filters": []
+          |}""".stripMargin
+      }
+    }
+  }
+
+  describe("ComplexRoute") {
+
+    val newRoute = NewComplexRoute(
+      matcher = Matcher(
+        pathMatcher = Some(RegexPathMatcher("/hello-*"))
+      )
+    )
+
+    it("should unmarshall the ComplexRoute") {
+      val complexRoute = """{
+                           |  "description": "this is a route",
+                           |  "activateAt": "2015-10-10T10:10:10",
+                           |  "id": 1,
+                           |  "createdAt": "2015-10-10T10:10:10",
+                           |  "route": {
+                           |    "matcher": {
+                           |      "pathMatcher": {
+                           |        "match": "/hello-*",
+                           |        "type": "REGEX"
+                           |      }
+                           |    }
+                           |  }
+                           |}""".stripMargin.parseJson.convertTo[ComplexRoute]
+      complexRoute should be {
+        ComplexRoute(1,
+          "this is a route",
+          newRoute,
+          LocalDateTime.of(2015, 10, 10, 10, 10, 10),
+          LocalDateTime.of(2015, 10, 10, 10, 10, 10),
+          None
+        )
+      }
+    }
+
+    it("should marshall the ComplexRoute") {
+
+      val routeJson = ComplexRoute(1,
+        "this is a route",
+        newRoute,
+        LocalDateTime.of(2015, 10, 10, 10, 10, 10),
+        LocalDateTime.of(2015, 10, 10, 10, 10, 10),
+        None
+      ).toJson
+
+      routeJson.prettyPrint should be {
+        """{
+          |  "description": "this is a route",
+          |  "activateAt": "2015-10-10T10:10:10",
+          |  "id": 1,
```

Comment:
if we don't have it numeric, this means the id will be generated on the client side... and we should enable put in innkeeper.

I feel that enabling put might be dangerous, as put means replacing a route with another one. And this will break our sync mechanism, as right now we query for routes which were created or deleted. Put combines a delete with a create.